--- a/com/mojang/realmsclient/exception/RetryCallException.java
+++ b/com/mojang/realmsclient/exception/RetryCallException.java
@@ -4,13 +4,13 @@
 {
     public final int field_224985_e;
 
-    public RetryCallException(int p_i242135_1_, int p_i242135_2_)
+    public RetryCallException(int p_i242136_1_, int p_i242136_2_)
     {
-        super(p_i242135_2_, "Retry operation", -1, "");
+        super(p_i242136_2_, "Retry operation", -1, "");
 
-        if (p_i242135_1_ >= 0 && p_i242135_1_ <= 120)
+        if (p_i242136_1_ >= 0 && p_i242136_1_ <= 120)
         {
-            this.field_224985_e = p_i242135_1_;
+            this.field_224985_e = p_i242136_1_;
         }
         else
         {
